Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/storage/CoronaTestStorage.kt

```diff
@@ -1,23 +1,64 @@
 package de.rki.coronawarnapp.coronatest.storage
 
+import android.content.Context
+import androidx.core.content.edit
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
 import de.rki.coronawarnapp.coronatest.type.CoronaTest
+import de.rki.coronawarnapp.coronatest.type.pcr.PCRCoronaTest
+import de.rki.coronawarnapp.coronatest.type.rapidantigen.RapidAntigenCoronaTest
+import de.rki.coronawarnapp.util.di.AppContext
+import de.rki.coronawarnapp.util.serialization.BaseGson
+import de.rki.coronawarnapp.util.serialization.adapter.RuntimeTypeAdapterFactory
 import timber.log.Timber
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
-class CoronaTestStorage @Inject constructor() {
+class CoronaTestStorage @Inject constructor(
+    @AppContext val context: Context,
+    @BaseGson val baseGson: Gson
+) {
 
-    suspend fun load(): Set<CoronaTest> {
-        Timber.tag(TAG).d("load()")
-        throw NotImplementedError()
+    private val prefs by lazy {
+        context.getSharedPreferences("coronatest_localdata", Context.MODE_PRIVATE)
     }
 
-    suspend fun save(tests: Set<CoronaTest>) {
-        Timber.tag(TAG).d("save(tests=%s)", tests)
+    private val gson by lazy {
+        baseGson.newBuilder().apply {
+            val rta = RuntimeTypeAdapterFactory.of(CoronaTest::class.java)
+                .registerSubtype(PCRCoronaTest::class.java)
+                .registerSubtype(RapidAntigenCoronaTest::class.java)
+
+            registerTypeAdapterFactory(rta)
+        }.create()
+    }
+
+    private val typeToken by lazy {
+        object : TypeToken<List<CoronaTest>>() {}.type
     }
 
+    var coronaTests: Collection<CoronaTest>
+        get() {
+            Timber.tag(TAG).d("load()")
+            val raw = prefs.getString(PKEY_TESTDATA, null) ?: return emptySet()
+            val tests: List<CoronaTest> = gson.fromJson(raw, typeToken)
+            tests.forEach {
+                Timber.tag(TAG).v("Loaded: %s", it)
+                requireNotNull(it.testGUID)
+            }
```

Comment:
I think you can print it directly , I assume it will crash anyway if the parser tried to assign nullable value to non nullable GUID property

Reply:
In certain edge cases it is possible to get invalid data, i.e. writing "empty data", and GSON will happily created a class instance and set non-nullable properties to null. This will only throw an exception when accessed.

Printing the object via Timber does not guarantee that the properties are evaluated, especially as the `it` will only be `it.toString()` if logging is actually enabled.

So the safer bet is to just access a specific attribute that should never be null.

In the previous case (I think exposure detection tracker), the issue of "empty data being deserialized" was likely caused writing to files directly and a buffer not being flushed. That's why I also used SharedPreference here, so we don't do direct file access. The previous issue was very very difficult to debug (https://github.com/corona-warn-app/cwa-app-android/pull/1715) due to our limited crashing tracking abilities, and was hotfixed in production.

To not be bitten by that again I'm doing the check here, to crash early, instead when the values are accessed later, which generates weird stacktraces.